File: saral_hr/saral_hr/report/provident_fund_register/provident_fund_register.js
// Copyright (c) 2026, sj and contributors
// For license information, please see license.txt

frappe.query_reports["Provident Fund Register"] = {
    filters: [
        {
            fieldname: "year",
            label: __("Year"),
            fieldtype: "Select",
            options: get_pf_year_options(),
            reqd: 1,
            default: ""
        },
        {
            fieldname: "month",
            label: __("Month"),
            fieldtype: "Select",
            options: [
                "", "January", "February", "March", "April",
                "May", "June", "July", "August",
                "September", "October", "November", "December"
            ],
            reqd: 1,
            default: ""
        },
        {
            fieldname: "company",
            label: __("Company"),
            fieldtype: "MultiSelectList",
            reqd: 1,
            get_data: function(txt) {
                return frappe.db.get_link_options("Company", txt);
            }
        },
        {
            fieldname: "category",
            label: __("Category"),
            fieldtype: "Link",
            options: "Category",
            reqd: 1,
            default: ""
        }
    ],

    onload: function(report) {
        frappe.after_ajax(() => {
            report.page.set_primary_action(__("Print"), () => {
                pf_register_print(report);
            }, "printer");
        });
    },

    formatter: function(value, row, column, data, default_formatter) {
        if (!data) return default_formatter(value, row, column, data);
        const fn = column.fieldname;

        if (data._row_type === "total") {
            if (["pf_no","uan_no","days","absent","date_of_joining","date_of_birth"].includes(fn)) return "";
            value = default_formatter(value, row, column, data);
            return `<strong>${value}</strong>`;
        }
        if ((fn === "vol_pf") && (value === null || value === undefined || value === "")) return "";
        return default_formatter(value, row, column, data);
    }
};

function get_pf_year_options() {
    let y = new Date().getFullYear();
    let years = [""];
    for (let i = y - 2; i <= y + 1; i++) years.push(i.toString());
    return years;
}

// ── Print ─────────────────────────────────────────────────────────────────────

function pf_register_print(report) {
    let data    = frappe.query_report.data || [];
    let filters = frappe.query_report.get_values() || {};
    let month   = filters.month || "";
    let year    = filters.year  || "";
    let company = frappe.boot.sysdefaults.company || "";

    let rows      = data.filter(r => r._row_type === "detail");
    let total_row = data.find(r => r._row_type === "total") || {};

    let tbody = rows.map((row, idx) => `
        <tr class="${idx % 2 !== 0 ? 'alt' : ''}">
            <td>${row.pf_no || ""}</td>
            <td>${row.uan_no || ""}</td>
            <td>${row.employee_id || ""}</td>
            <td>${row.employee_name || ""}</td>
            <td class="c">${row.days !== null ? row.days : ""}</td>
            <td class="c">${row.absent !== null ? row.absent : ""}</td>
            <td class="r">${fmt(row.gross)}</td>
            <td class="r">${fmt(row.basic_da)}</td>
            <td class="r">${fmt(row.emp_pf)}</td>
            <td class="r">${fmt(row.employer_eps)}</td>
            <td class="r">${fmt(row.employer_pf)}</td>
            <td class="r">${fmt(row.total_amount)}</td>
            <td class="r">${fmt(row.non_contrib)}</td>
            <td class="r">${row.vol_pf ? fmt(row.vol_pf) : ""}</td>
            <td class="r">${fmt(row.cumul_pf)}</td>
            <td class="r">${fmt(row.cumul_eps)}</td>
            <td class="r">${fmt(row.total_amount2)}</td>
            <td class="c">${row.date_of_joining || ""}</td>
            <td class="c">${row.date_of_birth || ""}</td>
        </tr>
    `).join("");

    tbody += `
        <tr class="total-row">
            <td colspan="6" class="total-label">Total</td>
            <td class="r">${fmt(total_row.gross)}</td>
            <td class="r">${fmt(total_row.basic_da)}</td>
            <td class="r">${fmt(total_row.emp_pf)}</td>
            <td class="r">${fmt(total_row.employer_eps)}</td>
            <td class="r">${fmt(total_row.employer_pf)}</td>
            <td class="r">${fmt(total_row.total_amount)}</td>
            <td class="r">${fmt(total_row.non_contrib)}</td>
            <td class="r">${total_row.vol_pf ? fmt(total_row.vol_pf) : ""}</td>
            <td class="r">${fmt(total_row.cumul_pf)}</td>
            <td class="r">${fmt(total_row.cumul_eps)}</td>
            <td class="r">${fmt(total_row.total_amount2)}</td>
            <td></td><td></td>
        </tr>
    `;

    let html = `<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Provident Fund Register – ${month} ${year}</title>
    <style>
        * { margin: 0; padding: 0; box-sizing: border-box; }
        body {
            font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", sans-serif;
            font-size: 8px;
            color: #333;
            background: #fff;
            -webkit-print-color-adjust: exact;
            print-color-adjust: exact;
        }
        .page {
            width: 420mm;
            min-height: 297mm;
            margin: 0 auto;
            padding: 10mm 12mm;
        }
        .header { text-align: center; margin-bottom: 12px; }
        .company-name { font-size: 13px; font-weight: 700; text-transform: uppercase; color: #1a1a1a; }
        .report-title { font-size: 11px; font-weight: 600; color: #333; margin-top: 3px; }
        .period { font-size: 9px; color: #888; margin-top: 2px; }

        hr.divider { border: none; border-top: 1px solid #d1d8dd; margin-bottom: 12px; }

        table {
            width: 100%;
            border-collapse: collapse;
            font-size: 8px;
            border: 1px solid #d1d8dd;
        }
        thead th {
            background: #f5f7fa;
            color: #555;
            font-weight: 600;
            font-size: 7.5px;
            text-transform: uppercase;
            letter-spacing: 0.2px;
            padding: 6px 5px;
            border-bottom: 1px solid #d1d8dd;
            border-right: 1px solid #ebecee;
            text-align: center;
            white-space: nowrap;
        }
        thead th:last-child { border-right: none; }
        thead th.l { text-align: left; }

        tbody td {
            padding: 5px 5px;
            color: #333;
            border-bottom: 1px solid #ebecee;
            border-right: 1px solid #ebecee;
        }
        tbody td:last-child { border-right: none; }
        tbody tr:last-child td { border-bottom: none; }
        tr.alt { background-color: #fafbfc; }

        .c { text-align: center; }
        .r { text-align: right; white-space: nowrap; }

        .total-row td {
            background: #f5f7fa;
            font-weight: 700;
            color: #222;
            border-top: 1px solid #d1d8dd;
            border-bottom: none;
            padding: 6px 5px;
        }
        .total-label { text-align: right; color: #555; font-weight: 600; }

        .footer { margin-top: 30px; display: flex; justify-content: space-between; }
        .sign { text-align: center; width: 130px; }
        .sign .line { border-top: 1px solid #bbb; margin-bottom: 5px; }
        .sign .lbl { font-size: 7.5px; color: #999; text-transform: uppercase; letter-spacing: 0.8px; }

        @media print {
            @page { size: A3 landscape; margin: 0; }
            .page { padding: 10mm 12mm; }
        }
    </style>
</head>
<body>
<div class="page">
    <div class="header">
        <div class="company-name">${company}</div>
        <div class="report-title">Provident Fund Register</div>
        <div class="period">For the Month of ${month} ${year}</div>
    </div>
    <hr class="divider">
    <table>
        <thead>
            <tr>
                <th class="l" style="width:70px;">PF No.</th>
                <th class="l" style="width:80px;">UAN No.</th>
                <th class="l" style="width:80px;">Employee ID</th>
                <th class="l" style="width:110px;">Employee Name</th>
                <th style="width:40px;">Days<br>(LWP+ABS)</th>
                <th style="width:38px;">Absent</th>
                <th style="width:80px;">Gross Salary</th>
                <th style="width:75px;">BS+DA<br>(AR+CO)</th>
                <th style="width:75px;">Emp PF<br>12%</th>
                <th style="width:78px;">Er EPS<br>8.33%</th>
                <th style="width:78px;">Er PF<br>3.67%</th>
                <th style="width:75px;">Total<br>Amount</th>
                <th style="width:75px;">Non<br>Cont.</th>
                <th style="width:55px;">Vol PF</th>
                <th style="width:72px;">Cumul PF</th>
                <th style="width:72px;">Cumul EPS</th>
                <th style="width:75px;">Total<br>Amount</th>
                <th style="width:72px;">Date of<br>Joining</th>
                <th style="width:72px;">Date of<br>Birth</th>
            </tr>
        </thead>
        <tbody>
            ${tbody}
        </tbody>
    </table>
    <div class="footer">
        <div class="sign"><div class="line"></div><div class="lbl">Prepared By</div></div>
        <div class="sign"><div class="line"></div><div class="lbl">Checked By</div></div>
        <div class="sign"><div class="line"></div><div class="lbl">Authorised Signatory</div></div>
    </div>
</div>
</body>
</html>`;

    let w = window.open("", "_blank");
    w.document.write(html);
    w.document.close();
    w.focus();
    setTimeout(() => { w.print(); }, 800);
}

function fmt(val) {
    if (val === null || val === undefined || val === "") return "";
    return parseFloat(val || 0).toLocaleString('en-IN', {
        minimumFractionDigits: 2,
        maximumFractionDigits: 2
    });
}

File: saral_hr/saral_hr/report/provident_fund_register/provident_fund_register.py
# Copyright (c) 2026, sj and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import flt
import math

MONTH_MAP = {
    'January': 1, 'February': 2, 'March': 3, 'April': 4,
    'May': 5, 'June': 6, 'July': 7, 'August': 8,
    'September': 9, 'October': 10, 'November': 11, 'December': 12
}

PF_WAGE_CEILING = 15000.0


def execute(filters=None):
    filters = filters or {}
    validate_filters(filters)
    columns = get_columns()
    data    = get_data(filters)
    return columns, data


def validate_filters(filters):
    if not filters.get("month"):
        frappe.throw("Please select a Month")
    if not filters.get("year"):
        frappe.throw("Please select a Year")
    if not filters.get("company"):
        frappe.throw("Please select at least one Company")
    if not filters.get("category"):
        frappe.throw("Please select a Category")


def get_columns():
    return [
        {"label": "PF No.",             "fieldname": "pf_no",          "fieldtype": "Data",     "width": 110},
        {"label": "UAN No.",            "fieldname": "uan_no",          "fieldtype": "Data",     "width": 110},
        {"label": "Employee ID",        "fieldname": "employee_id",    "fieldtype": "Data",     "width": 120},
        {"label": "Employee Name",      "fieldname": "employee_name",  "fieldtype": "Data",     "width": 180},
        {"label": "Days (LWP+ABS)",     "fieldname": "days",           "fieldtype": "Float",    "width": 80,  "precision": 1},
        {"label": "Absent",             "fieldname": "absent",         "fieldtype": "Float",    "width": 65,  "precision": 1},
        {"label": "Gross Salary",       "fieldname": "gross",          "fieldtype": "Float", "width": 110, "precision": 2},
        {"label": "BS+DA (AR+CO)",      "fieldname": "basic_da",       "fieldtype": "Float", "width": 110, "precision": 2},
        {"label": "Employee PF 12%",    "fieldname": "emp_pf",         "fieldtype": "Float", "width": 110, "precision": 2},
        {"label": "Employer EPS 8.33%", "fieldname": "employer_eps",   "fieldtype": "Float", "width": 120, "precision": 2},
        {"label": "Employer PF 3.67%",  "fieldname": "employer_pf",    "fieldtype": "Float", "width": 120, "precision": 2},
        {"label": "Total Amount",       "fieldname": "total_amount",   "fieldtype": "Float", "width": 105, "precision": 2},
        {"label": "Non-Cont.",          "fieldname": "non_contrib",    "fieldtype": "Float", "width": 100, "precision": 2},
        {"label": "Vol. PF",            "fieldname": "vol_pf",         "fieldtype": "Float", "width": 80,  "precision": 2},
        {"label": "Cumul. PF",          "fieldname": "cumul_pf",       "fieldtype": "Float", "width": 95,  "precision": 2},
        {"label": "Cumul. EPS",         "fieldname": "cumul_eps",      "fieldtype": "Float", "width": 95,  "precision": 2},
        {"label": "Total Amount",       "fieldname": "total_amount2",  "fieldtype": "Float", "width": 105, "precision": 2},
        {"label": "Date of Joining",    "fieldname": "date_of_joining","fieldtype": "Date",     "width": 105},
        {"label": "Date of Birth",      "fieldname": "date_of_birth",  "fieldtype": "Date",     "width": 100},
    ]


def get_data(filters):
    import json

    month          = filters.get("month")
    year           = filters.get("year")
    category       = filters.get("category")       # single string (Link field)
    company_filter = filters.get("company")

    month_num  = MONTH_MAP.get(month)
    start_date = "%s-%02d-01" % (year, month_num)

    where_clauses = ["ss.docstatus = 1", "ss.start_date = %(start_date)s"]
    query_params  = {"start_date": start_date}

    if company_filter:
        if isinstance(company_filter, str):
            company_filter = json.loads(company_filter)
        if company_filter:
            where_clauses.append("ss.company IN %(companies)s")
            query_params["companies"] = tuple(company_filter)

    # Category filter via Company Link (same pattern as all other reports)
    category_join = ""
    if category:
        category_join = """
            INNER JOIN `tabCompany Link` cl
                ON  cl.name     = ss.employee
                AND cl.category = %(category)s
        """
        query_params["category"] = category

    where_sql = " AND ".join(where_clauses)

    # ── Salary slip base data ──────────────────────────────────────────────
    slips = frappe.db.sql("""
        SELECT
            ss.name           AS slip_name,
            ss.employee       AS employee_id,
            ss.employee_name  AS employee_name,
            ss.payment_days   AS days,
            ss.absent_days    AS absent,
            ss.total_earnings AS gross
        FROM `tabSalary Slip` ss
        {category_join}
        WHERE {where}
        ORDER BY ss.employee_name
    """.format(where=where_sql, category_join=category_join), query_params, as_dict=1)

    if not slips:
        return []

    slip_names = tuple(s.slip_name for s in slips)

    # ── Basic + DA from earnings ───────────────────────────────────────────
    basic_da_rows = frappe.db.sql("""
        SELECT
            sd.parent AS slip_name,
            SUM(sd.amount) AS basic_da
        FROM `tabSalary Details` sd
        WHERE sd.parent IN %(slip_names)s
          AND sd.parenttype  = 'Salary Slip'
          AND sd.parentfield = 'earnings'
          AND (
              LOWER(sd.salary_component) LIKE %(like_basic)s
              OR LOWER(sd.salary_component) LIKE %(like_da)s
              OR LOWER(sd.salary_component) LIKE %(like_dear)s
          )
        GROUP BY sd.parent
    """, {
        "slip_names": slip_names,
        "like_basic": "%basic%",
        "like_da":    "%dearness%",
        "like_dear":  "% da",
    }, as_dict=1)

    basic_da_map = {r.slip_name: flt(r.basic_da) for r in basic_da_rows}

    # ── Voluntary PF from deductions ──────────────────────────────────────
    vpf_rows = frappe.db.sql("""
        SELECT
            sd.parent AS slip_name,
            SUM(sd.amount) AS vol_pf
        FROM `tabSalary Details` sd
        WHERE sd.parent IN %(slip_names)s
          AND sd.parenttype  = 'Salary Slip'
          AND sd.parentfield = 'deductions'
          AND LOWER(sd.salary_component) LIKE %(like_vpf)s
        GROUP BY sd.parent
    """, {
        "slip_names": slip_names,
        "like_vpf":   "%voluntary%",
    }, as_dict=1)

    vpf_map = {r.slip_name: flt(r.vol_pf) for r in vpf_rows}

    # ── Employee master: PF No, UAN, DOJ, DOB ─────────────────────────────
    employee_ids = tuple(s.employee_id for s in slips)

    emp_rows = frappe.db.sql("""
        SELECT
            e.name,
            e.employee_pf_account  AS pf_no,
            e.pf_uan_number        AS uan_no,
            e.date_of_birth,
            cl.date_of_joining
        FROM `tabEmployee` e
        LEFT JOIN `tabCompany Link` cl ON cl.name = e.name
        WHERE e.name IN %(employee_ids)s
    """, {"employee_ids": employee_ids}, as_dict=1)

    emp_map = {r.name: r for r in emp_rows}

    # ── Build rows ─────────────────────────────────────────────────────────
    data           = []
    t_gross        = 0.0
    t_basic_da     = 0.0
    t_emp_pf       = 0.0
    t_er_eps       = 0.0
    t_er_pf        = 0.0
    t_total        = 0.0
    t_nc           = 0.0
    t_vpf          = 0.0
    t_cumul_pf     = 0.0
    t_cumul_eps    = 0.0
    t_total2       = 0.0

    for idx, slip in enumerate(slips, start=1):
        emp        = emp_map.get(slip.employee_id, frappe._dict())
        gross      = flt(slip.gross, 2)
        actual_bda = flt(basic_da_map.get(slip.slip_name, 0.0), 2)

        # PF wage capped at 15,000
        pf_wage    = min(actual_bda, PF_WAGE_CEILING)

        # Employee PF = 12% of pf_wage (rounded)
        emp_pf     = flt(round(pf_wage * 0.12), 2)

        # Employer EPS = round(emp_pf × 0.6944)  [8.33/12]
        er_eps     = flt(round(emp_pf * 0.6944), 2)

        # Employer PF diff = round(emp_pf × 0.3056)  [3.67/12]
        er_pf      = flt(round(emp_pf * 0.3056), 2)

        # Total = emp_pf + emp_pf
        total_amt  = flt(emp_pf + emp_pf, 2)

        # Non-Contributory = max(0, actual BS+DA - 15000)
        non_contrib = flt(max(0.0, actual_bda - PF_WAGE_CEILING), 2)

        # Voluntary PF
        vol_pf     = flt(vpf_map.get(slip.slip_name, 0.0), 2)

        # Cumul PF = emp_pf + er_pf (current month only)
        cumul_pf   = flt(emp_pf + er_pf, 2)

        # Cumul EPS = er_eps (current month only)
        cumul_eps  = flt(er_eps, 2)

        # Total Amount col 2 = same as total_amt
        total_amt2 = total_amt

        t_gross     += gross
        t_basic_da  += actual_bda
        t_emp_pf    += emp_pf
        t_er_eps    += er_eps
        t_er_pf     += er_pf
        t_total     += total_amt
        t_nc        += non_contrib
        t_vpf       += vol_pf
        t_cumul_pf  += cumul_pf
        t_cumul_eps += cumul_eps
        t_total2    += total_amt2

        data.append({
            "pf_no":          emp.get("pf_no") or "",
            "uan_no":         emp.get("uan_no") or "",
            "employee_id":    slip.employee_id,
            "employee_name":  slip.employee_name,
            "days":           flt(slip.days, 1),
            "absent":         flt(slip.absent, 1),
            "gross":          gross,
            "basic_da":       actual_bda,
            "emp_pf":         emp_pf,
            "employer_eps":   er_eps,
            "employer_pf":    er_pf,
            "total_amount":   total_amt,
            "non_contrib":    non_contrib,
            "vol_pf":         vol_pf if vol_pf else None,
            "cumul_pf":       cumul_pf,
            "cumul_eps":      cumul_eps,
            "total_amount2":  total_amt2,
            "date_of_joining":emp.get("date_of_joining") or "",
            "date_of_birth":  emp.get("date_of_birth") or "",
            "_row_type":      "detail",
        })

    # Grand total
    data.append({
        "pf_no":          "",
        "uan_no":         "",
        "employee_id":    "",
        "employee_name":  "Total",
        "days":           None,
        "absent":         None,
        "gross":          flt(t_gross, 2),
        "basic_da":       flt(t_basic_da, 2),
        "emp_pf":         flt(t_emp_pf, 2),
        "employer_eps":   flt(t_er_eps, 2),
        "employer_pf":    flt(t_er_pf, 2),
        "total_amount":   flt(t_total, 2),
        "non_contrib":    flt(t_nc, 2),
        "vol_pf":         flt(t_vpf, 2) if t_vpf else None,
        "cumul_pf":       flt(t_cumul_pf, 2),
        "cumul_eps":      flt(t_cumul_eps, 2),
        "total_amount2":  flt(t_total2, 2),
        "date_of_joining": "",
        "date_of_birth":   "",
        "bold":            1,
        "_row_type":       "total",
    })

    return data


# Here are the column calculations in clear points:
# 1. Employee PF (12%)

# round(pf_wage × 0.12)
# where pf_wage = min(actual BS+DA, 15,000)

# 2. Employer EPS (8.33%)

# round(emp_pf × 0.6944)
# 0.6944 = 8.33 ÷ 12

# 3. Employer PF (3.67%)

# round(emp_pf × 0.3056)
# 0.3056 = 3.67 ÷ 12

# 4. Total Amount

# emp_pf + emp_pf
# Employee contribution + Employer total (both sides equal emp_pf)

# 5. Non-Contributory

# max(0, actual BS+DA − 15,000)
# Only the salary portion above the ₹15,000 PF ceiling

# 6. Voluntary PF

# Actual amount of deduction component named like %voluntary% from the salary slip

# 7. Cumul. PF

# emp_pf + employer_pf (current month only, not a running total across months)

# 8. Cumul. EPS

# employer_eps (current month only)

# 9. Total Amount (2nd col)

# Same as Total Amount — emp_pf + emp_pf